src/Riok.Mapperly/Descriptors/Enumerables/CollectionInfoBuilder.cs

```diff
@@ -188,21 +188,147 @@ or CollectionType.ReadOnlyMemory
         return null;
     }
 
-    private static CollectionType GetImplementedCollectionTypes(WellKnownTypes types, ITypeSymbol type)
+    private static CollectionType GetImplementedCollectionTypes(WellKnownTypes types, ITypeSymbol type, CollectionType collectionType)
     {
-        var implementedCollectionTypes = type.IsArrayType() ? CollectionType.Array : CollectionType.None;
-
-        foreach (var typeInfo in _collectionTypeInfos)
+        return collectionType switch
```

Comment:
Could you add a comment that the dynamic fallback `IterateHierarchy` is only used, if the type is not well-known and that this is done for performance reasons.

Reply:
👍  Changed name to `IterateImplementedTypes` and added comments